File: lc/c/009.c

static int compare(const void *a, const void *b) {
    return *(int*)a - *(int*)b;
}

static void dfs(int *combo, int comboSize, int total, int index){
    if (total == target) {
        
    }
}
/**
 * Return an array of arrays of size *returnSize.
 * The sizes of the arrays are returned as *returnColumnSizes array.
 * Note: Both returned array and *columnSizes array must be malloced, assume caller calls free().
 */
int** combinationSum(int* candidates, int candidatesSize, int target, int* returnSize, int** returnColumnSizes){
    int max_sz = 64;
    int **ret = malloc(max_sz * sizeof *ret);
    int sz = 0;

    qsort(candidates, candidatesSize, sizeof *candidates, compare);

}

File: lc/py/009.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        def dfs(combo, total, index):
            if total == target:
                ret.append(combo[:])
                return

            for i in range(index, len(candidates)):
                if total+candidates[i] > target:
                    return
                combo.append(candidates[i])
                dfs(combo, total+candidates[i], i)
                combo.pop()

        ret = []
        candidates.sort()
        dfs([], 0, 0)
        return ret

# slow
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        def dfs(combo, total, index):
            if total > target:
                return
            elif total == target:
                ret.append(combo[:])
                return

            for i in range(index, len(candidates)):
                combo.append(candidates[i])
                dfs(combo, total+candidates[i], i)
                combo.pop()

        ret = []
        dfs([], 0, 0)
        return ret

# subtract target
def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        result = []

        def dfs(combo, total, index):
            if total == 0:
                result.append(combo[:])
                return
            for i in range(index, len(candidates)):
                if total - candidates[i] < 0:
                    return
                combo.append(candidates[i])
                dfs(combo, total-candidates[i], i)
                combo.pop()

        candidates.sort()
        dfs([], target, 0)
        return result

# slow
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        result = []
        def dfs(total, index):
            if total == 0:
                result.append(combo[:])
            elif total > 0:
                for i in range(index, len(candidates)):
                    combo.append(candidates[i])
                    dfs(total-candidates[i], i)
                    combo.pop()

        candidates.sort()
        combo = []
        dfs(target, 0)
        return result
